cnd/src/ethereum.rs

```diff
@@ -1,22 +1,140 @@
 #![warn(unused_extern_crates, missing_debug_implementations, rust_2018_idioms)]
 #![forbid(unsafe_code)]
 
+use bitcoin::hashes::core::{fmt::Formatter, str::FromStr};
+use core::fmt;
 pub use ethbloom::{Bloom as H2048, Input};
-pub use primitive_types::{H160, H256, U128, U256};
+use hex::FromHexError;
+pub use primitive_types::U256;
 use serde::{Deserialize, Serialize};
 use serde_hex::{CompactPfx, SerHex, SerHexSeq, StrictPfx};
-
-pub type Address = H160;
+use std::fmt::Display;
 
 #[derive(Debug, Default, Copy, Clone, PartialEq, Deserialize)]
 pub struct H64(#[serde(with = "SerHex::<StrictPfx>")] [u8; 8]);
 
+#[derive(
+    Debug, Default, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize,
+)]
+pub struct Address(#[serde(with = "SerHex::<StrictPfx>")] [u8; 20]);
+
+impl Address {
+    pub fn from_slice(src: &[u8]) -> Self {
+        let mut address = Address([0u8; 20]);
+        address.0.copy_from_slice(src);
+        address
+    }
+}
+
+impl From<[u8; 20]> for Address {
+    fn from(bytes: [u8; 20]) -> Self {
+        Address(bytes)
+    }
+}
+
+impl From<Address> for [u8; 20] {
+    fn from(s: Address) -> Self {
+        s.0
+    }
+}
+
+impl FromStr for Address {
+    type Err = FromHexError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        hex::decode(s).map(|v| Address::from_slice(v.as_slice()))
+    }
+}
+
+impl From<Address> for Hash {
+    fn from(address: Address) -> Self {
+        let mut h256 = Hash([0u8; 32]);
+        h256.0[(32 - 20)..32].copy_from_slice(&address.0);
+        h256
+    }
+}
+
+impl fmt::LowerHex for Address {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        if f.alternate() {
+            write!(f, "0x")?;
+        }
+        for i in &self.0[..] {
+            write!(f, "{:02x}", i)?;
+        }
+        Ok(())
+    }
+}
+
+#[derive(
+    Debug, Default, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize,
+)]
+pub struct Hash(#[serde(with = "SerHex::<StrictPfx>")] [u8; 32]);
+
+impl From<[u8; 32]> for Hash {
+    fn from(bytes: [u8; 32]) -> Self {
+        Hash(bytes)
+    }
+}
+
+impl From<Hash> for [u8; 32] {
+    fn from(s: Hash) -> Self {
+        s.0
+    }
+}
+
+impl Hash {
+    pub fn from_slice(src: &[u8]) -> Self {
+        let mut h256 = Hash([0u8; 32]);
+        h256.0.copy_from_slice(src);
+        h256
+    }
+
+    pub fn as_bytes(&self) -> &[u8] {
+        &self.0
+    }
+}
+
+impl FromStr for Hash {
+    type Err = FromHexError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        hex::decode(s).map(|v| Hash::from_slice(v.as_slice()))
+    }
+}
+
+impl fmt::LowerHex for Hash {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        if f.alternate() {
+            write!(f, "0x")?;
+        }
+        for i in &self.0[..] {
+            write!(f, "{:02x}", i)?;
+        }
+        Ok(())
+    }
+}
+
+impl Display for Hash {
```

Comment:
Eventually, I would like this `Display` implementation to not print `...` but do what the implementation of `LowerHex` does.
We can then remove the implementation of `LowerHex`.

This will likely require a few changes to log statements in our code where we currently use `{:x}` to print the full transaction hash.

Can be done as a follow-up ticket :)

Reply:
recorded: https://github.com/comit-network/comit-rs/issues/2413